R/knitr-engine.R

```diff
@@ -628,3 +642,54 @@ eng_python_autoprint <- function(captured, options, autoshow) {
   }
 
 }
+
+save_python_session <- function(cache_path) {
+  module <- tryCatch(import("dill"), error = identity)
+  if (inherits(module, "error")) {
+    if (module$message == "ImportError: No module named dill") return()
+    signalCondition(module$message)
+  }
+
+  r_obj_exists <- "'r' in globals()"
+  r_is_R <- "type(r).__module__ == '__main__' and type(r).__name__ == 'R'"
+  if (py_eval(r_obj_exists) && py_eval(r_is_R)) {
+    py_run_string("del globals()['r']")
```

Comment:
Wouldn't this have side effects (basically meaning the `r` object is no longer visible after this code is run)?

Reply:
This is run at the knitr block end, and currently the `r` object in injected again in the Python namespace at the beginning of the following block. An alternative to putting this object directly in the user's `__main__` namespace would be to add it to `__builtins__`.

Reply:
Didn't put the `r` object in `__builtins__`, but I did put the "R object class" in `__builtins__`. The `r` object is not removed before saving the cache anymore, it's just ignored. However, my previous suggestion is still open for discussion.